src/extractor/guidance/coordinate_extractor.cpp

```diff
@@ -0,0 +1,834 @@
+                                             const EdgeID turn_edge,
+                                             const bool traversed_in_reverse,
+                                             const NodeID to_node) const
+{
+    if (!compressed_geometries.HasEntryForID(turn_edge))
+    {
+        if (traversed_in_reverse)
+            return {{node_coordinates[to_node]}, {node_coordinates[intersection_node]}};
+        else
+            return {{node_coordinates[intersection_node]}, {node_coordinates[to_node]}};
+    }
+    else
+    {
+        // extracts the geometry in coordinates from the compressed edge container
+        std::vector<util::Coordinate> result;
+        const auto &geometry = compressed_geometries.GetBucketReference(turn_edge);
+        result.reserve(geometry.size() + 2);
+
+        // the compressed edges contain node ids, we transfer them to coordinates accessing the
+        // node_coordinates array
+        const auto compressedGeometryToCoordinate = [this](
+            const CompressedEdgeContainer::OnewayCompressedEdge &compressed_edge) -> util::Coordinate {
+            return node_coordinates[compressed_edge.node_id];
+        };
+
+        // add the coordinates to the result in either normal or reversed order, based on
+        // traversed_in_reverse
+        if (traversed_in_reverse)
+        {
+            std::transform(geometry.rbegin(),
+                           geometry.rend(),
+                           std::back_inserter(result),
+                           compressedGeometryToCoordinate);
+            result.push_back(node_coordinates[intersection_node]);
+        }
+        else
+        {
+            result.push_back(node_coordinates[intersection_node]);
+            std::transform(geometry.begin(),
+                           geometry.end(),
+                           std::back_inserter(result),
+                           compressedGeometryToCoordinate);
+        }
+        return result;
+    }
+}
+
+double
+CoordinateExtractor::GetMaxDeviation(std::vector<util::Coordinate>::const_iterator range_begin,
+                                     const std::vector<util::Coordinate>::const_iterator &range_end,
+                                     const util::Coordinate &straight_begin,
+                                     const util::Coordinate &straight_end) const
+{
+    double deviation = 0;
+    for (; range_begin != range_end; range_begin = std::next(range_begin))
+    {
+        // find the projected coordinate
+        auto coord_between = util::coordinate_calculation::projectPointOnSegment(
+                                 straight_begin, straight_end, *range_begin)
+                                 .second;
+        // and calculate the distance between the intermediate coordinate and the coordinate
+        // on the osrm-way
+        deviation =
+            std::max(deviation,
+                     util::coordinate_calculation::haversineDistance(coord_between, *range_begin));
+    }
```

Comment:
I take it all back, using accumulate with an additional comment to avoid potential misunderstandings of the concept.